lib/Transform/XTenQDQConcat.cpp

```diff
@@ -0,0 +1,76 @@
+#include "xten/Dialect/XTenNN/IR/XTenNNOps.h"
+#include "xten/Transform/Passes.h"
+
+#include <mlir/Dialect/Tosa/IR/TosaOps.h>
+#include <mlir/IR/Matchers.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/Support/LogicalResult.h>
+
+#include <llvm/ADT/SmallVector.h>
+
+using namespace mlir;
+using namespace mlir::tosa;
+using namespace amd::xten_nn;
+
+namespace {
+struct XTenQDQConcat : public OpRewritePattern<DequantizeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(DequantizeOp op,
+                                PatternRewriter &rewriter) const override {
+    // Match concat->QDQ->concat and remove QDQ, if concats would be foldable.
+    // Removing a QDQ is already destructive. Try to be a little-less
+    // destructive by checking that the QDQ nodes have the same shift.
+    auto quantize =
+        llvm::dyn_cast_or_null<QuantizeOp>(op.getInput().getDefiningOp());
+    if (!quantize) {
+      return rewriter.notifyMatchFailure(
+          op, "DequantizeOp input not produced by QuantizeOp.");
+    }
+
+    if (quantize.getShift() != op.getShift()) {
+      return rewriter.notifyMatchFailure(
+          op, "DequantizeOp and QuantizeOp do not share the same shift value.");
+    }
+
+    // Try to match an incoming concat
+    auto producer =
+        dyn_cast_or_null<ConcatOp>(quantize.getInput().getDefiningOp());
+    if (producer == nullptr) {
+      return rewriter.notifyMatchFailure(
+          op, "QDQ input not produced by TOSA concat.");
+    }
+
+    uint64_t axis = producer.getAxis();
+
+    bool modified = false;
+    for (Operation *user : op->getUsers()) {
+      auto userConcat = dyn_cast<ConcatOp>(user);
+      if (userConcat == nullptr)
+        continue;
+
+      // Concats are foldable if they concatenate on the same axis
+      if (userConcat.getAxis() != axis)
+        continue;
+
+      // Foldable concats found. Folding the concats is the responsibility of
+      // TOSA canonicalization. Rewire the concat->QDQ->concat to be
+      // concat->concat instead.
+      rewriter.startRootUpdate(user);
+      user->replaceUsesOfWith(op, producer.getOutput());
+      rewriter.finalizeRootUpdate(user);
+      modified = true;
+    }
+
+    if (modified)
+      return success();
+
+    return rewriter.notifyMatchFailure(
+        op, "No foldable concats found around QDQ node.");
+  }
+};
+} // namespace
+
+void xilinx::xten::populateXTenQDQConcatPatterns(RewritePatternSet &patterns) {
```

Comment:
Will we have more than one pattern here? Otherwise, I would argue we don't need to have `populateXTenQDQConcatPatterns` but just the single pattern added in `populateXTenDestructiveCanonicalizePatterns`.

Reply:
We will probably also have wrapping of maxpool/reshape with QDQs here as well in the `populateXTenQDQConcatPatterns`.

Reply:
The idea is two fold:
1. You can add more QDQConcat patterns in the future
2. It allows you to use these patterns without having to use all of the destructive patterns